### ISSUE TITLE
Set user-provided tags on all created sub-resources

### ISSUE THREAD

**Issue by ringods** [pulumi/pulumi-awsx#383]
Pulumi Crosswalk for AWS provides a set of components, which are a composite of the individual `pulumi-aws` resources. For a composite component, please make sure that user-provided AWS tags are set on all the contained individual components.

For instance, an `awsx.ec2.Vpc` (possibly) creates a VPC, subnets, routing tables, internet gateways, NAT gateways, ... The tags passed into the parent component only seem to be set on the VPC at the moment, not on the subnets and other resources.

For the VPC example, setting specific tags can be required for Kubernetes as was [asked by Scott Lowe](https://pulumi-community.slack.com/archives/CJ909TL6P/p1565385393089500) on the Pulumi Community Slack.

For AWS Cost Reporting based on tags, all resources should receive the tags. To report on the cost of a subset of resources within an AWS account, I use the tags `project` and `environment` consistently. Now I can see what my `development`, `staging` and `production` environments cost me per `project`.

Context:
* Could be a duplicate of #209, but that ticket only contains a title.
* Related to #161, #306 & #322

**Comment by CyrusNajmabadi** [comment]
@ringods This is indeed an area we want to create a solution for.  However, we don't necessarily think the right approach is to just take the tags and apply to all sub-resources.  It may be about providing a more flexibly approach where you get the chance to participate in resource construction and can decide on a per-resource basis precisely what tweaks you want to make. 

We'll let you now as we start work in this area to ensure any solution will be viable for you :)

**Comment by ringods** [comment]
@CyrusNajmabadi having a (for me) standard set of tags on all resources is all I need for AWS cost reporting. So please do not over-engineer this. My use case is simple.

**Comment by CyrusNajmabadi** [comment]
@ringods Thanks for clarifying your use case.  We def want to make sure that you can have a solution that works for you.  However, we also need to be cognizant of feedback we've heard from otehrs where tagging is important and it's *not* the case that all teh resources get the exact same set of tags.  It would be a pity to add a solutoin only to have it have to be replaced with a different solution later to meet the needs of others in a similar, but not exactly the same, situation as you.  Thanks! :)

**Comment by seivan** [comment]
I agree with @ringods just simply tagging the stuff Pulumi creates using some prefix based on stack would be very useful so we can see the resources on the AWS console. Don't really need more than that.

**Comment by ctwise** [comment]
Would love to have this standardized capability. Minimally provide one or more tags to apply to everything, maximally provide a callback that can create tags based on the resource.

**Comment by stewartnoll** [comment]
Although it's a bit of a hack while we wait on this to be standardized, here's a list of all AWS resources configurable by Pulumi that support tags: https://gist.github.com/stewartnoll/bc8aacff9e785be5c12db320e2852f5f

And to wire it up:
```
import { runtime } from "@pulumi/pulumi";
import { resourcesWithTags } from './awsResourcesWithTags';

runtime.registerStackTransformation(args => {
    if (resourcesWithTags.indexOf(args.type) !== -1) {
        return {
            ...args,
            props: {
                ...args.props,
                tags: {
                    owner: "stewartnoll"
                }
            }
        }
    }
    return undefined;
});

...
```

**Comment by ypresto** [comment]
@stewartnoll FYI: It's a large list in a heavily called function. Perhaps `new Set(resourcesWithTags)` improves performance.

**Comment by stewartnoll** [comment]
post from @joeduffy utilizing a similar approach as above... https://www.pulumi.com/blog/automatically-enforcing-aws-resource-tagging-policies/#automatically-applying-tags

**Comment by damianh** [comment]
I would like to add this is causing us issues with our policies-as-code (enforcement of tags). We're going to apply the `registerAutoTags` from the above blog post, but it feels hacky and less cohesive.

**Comment by ctwise** [comment]
Totally agree. Wish this was baked in. We have to have this code in every single stack.

**Comment by martinek-stepan** [comment]
Is there anyway to generate that list? Because it is missing for example all apigatewayv2 resources and pulumi already know what resources allow tagging as you can specify it only for those resources in argument when creating them.

**Comment by paulrobu** [comment]
> 
Earlier https://github.com/pulumi/pulumi-awsx/issues/383#issuecomment-577963947 points to https://gist.github.com/stewartnoll/bc8aacff9e785be5c12db320e2852f5f which shows a way to obtain an up to date list of resources based on https://github.com/pulumi/pulumi-aws/tree/master/sdk/nodejs repo:
```
/**
 * A list of all AWS resources that support tags.
 * 
 * Generated with the help of:
 * grep -rl 'public readonly tags!' <path-to-pulumi-aws-repo>/sdk/nodejs/ 
 * | xargs grep 'public static readonly __pulumiType' 
 * | awk '{print $7}'
 * | sed 's/.$/,/'
*/
```
I saved the list of AWS taggable resources as of today at https://gist.github.com/paulrobu/93deb7a54e1e3cc43ce255ea62d9288b

**Comment by ringods** [comment]
I made a code example on using a resource transformation to set tags when using the AWS native provider:

https://github.com/ringods/pulumi-samples/tree/main/aws-native-tagging

Adding this here in case someone is looking at a solution for this and bumping into this issue.

**Comment by mnlumi** [comment]
I tested with awsx 2.1.1 and vpc components.This looks to be resolved for both defaultTags at the provider level as well as resource tags. Both are propagating down to all the resources below it with the exception of EIP, which I've opened a separate PR for here: https://github.com/pulumi/pulumi-awsx/pull/1155. 

@ringods Please let me know if awsx 2.1.1 matches expected behavior now (with the exception of EIP).
